### PR TITLE
8310130: C2: assert(false) failed: scalar_input is neither phi nor a matchin reduction

### PR DESCRIPTION
**Pull request by eme64**
Removed a spurious assert before optimization bailout.

I assumed that the "scalar input" of a Reduction node must always be either a Phi or another Reduction node. But that is incorrect, partial vectorization can lead to a reduction node chain where we have vector reductions and scalar reductions at the same time. In those cases, we cannot move the UnorderedReductions out of the loop, so a optimization bailout is appropriate.

I assessed the other asserts in `PhaseIdealLoop::move_unordered_reduction_out_of_loop`, and I think they are all justified. However, one assert would have lead to a `continue` in production, which would not break out of the nested loop correctly. I changed it to a `return`, so that would be a bailout from the optimization. This assert should not be triggered because in `SuperWord::mark_reductions` we forbid that a reduction node has any uses inside the loop except for the successor node in the reduction chain.

I have one regression test delivered by the fuzzer, and one that I constructed myself after understanding the issue.
Testing up to tier6 and stress testing.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8310130](https://bugs.openjdk.org/browse/JDK-8310130): C2: assert(false) failed: scalar_input is neither phi nor a matchin reduction (**Bug** - P3)

### Reviewers
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**)
 * [Christian Hagedorn](https://openjdk.org/census#chagedorn) (@chhagedorn - **Reviewer**) ⚠️ Review applies to [53e8913e](https://git.openjdk.org/jdk/pull/14494/files/53e8913ec04a7a466ef8e0fc68c8d78bdecf80f9)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/14494/head:pull/14494` \
`$ git checkout pull/14494`

Update a local copy of the PR: \
`$ git checkout pull/14494` \
`$ git pull https://git.openjdk.org/jdk.git pull/14494/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 14494`

View PR using the GUI difftool: \
`$ git pr show -t 14494`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/14494.diff">https://git.openjdk.org/jdk/pull/14494.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/14494#issuecomment-1596778249)

### CODE REVIEW

**Review by vnkozlov — APPROVED** [review]
Looks fine.

File: test/hotspot/jtreg/compiler/loopopts/superword/TestUnorderedReductionPartialVectorization.java [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+
+import compiler.lib.ir_framework.*;
+
+public class TestUnorderedReductionPartialVectorization {
+    static final int RANGE = 1024;
+    static final int ITER  = 10;
+
+    public static void main(String[] args) {
+        TestFramework.runWithFlags("-Xbatch", "-XX:-TieredCompilation",
+                                   "-XX:CompileCommand=compileonly,compiler.loopopts.superword.TestUnorderedReductionPartialVectorization::test*");
```
Comment (chhagedorn) [inline test/hotspot/jtreg/compiler/loopopts/superword/TestUnorderedReductionPartialVectorization.java:43]: It should probably also trigger without the flags by just specifying `TestFramework.run()` as `test1()` is not using other methods and the IR framework will implicitly use `-Xbatch` and wait for the compilation of `test1()` to be finished.

**Review by chhagedorn — APPROVED** [review]
Looks good!

**Review by chhagedorn — APPROVED** [review]

File: test/hotspot/jtreg/compiler/loopopts/superword/TestUnorderedReductionPartialVectorization.java (hunk 1 above)
Comment (chhagedorn) [reply test/hotspot/jtreg/compiler/loopopts/superword/TestUnorderedReductionPartialVectorization.java]: Nice!

**Comment by eme64** [comment]
Just saw a x86 (32bit) IR rule failure for `compiler.loopopts.superword.TestUnorderedReductionPartialVectorization.test1`. Will investigate.

**Review by vnkozlov — APPROVED** [review]
Update is good

**Comment by eme64** [comment]
Thanks @vnkozlov @chhagedorn for the reviews!
/integrate

**Pull request closed without merging**